### PR TITLE
py-pyhamcrest: new port as a dependency of py-twisted

### PR DESCRIPTION
**Pull request by mojca**
#### Description

While trying to install buildbot slave I discovered a problem that prevented me from running the build slave:
```
pkg_resources.DistributionNotFound: The 'PyHamcrest>=1.9.0' distribution was not found and is required by Twisted
```

See: https://github.com/twisted/twisted/blob/trunk/src/twisted/python/_setup.py#L241

Questions:
* Not sure about what the best name should be for pyhamcrest. The upstream name is `PyHamcrest`.
* Not sure whether to remove 3.4 support from `py-twisted` or to add it to `py-hamcrest`. Other dependencies are missing in order to be able to add `py37-twisted`.
* There might be other dependencies missing (automat, attrs).

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.13.6 17G65
Xcode 9.4.1 9F2000 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by ryandesign** [comment]
> Not sure about what the best name should be for pyhamcrest. The upstream name is PyHamcrest.

IIRC there was a concerted effort a few years back to rename MacPorts Python modules to "py-" followed by the upstream name, even if the upstream name begins with "py". So that would make this "py-PyHamcrest" or "py-pyhamcrest". (I'm not sure which capitalization is preferred.)

**Comment by mojca** [comment]
OK, thanks, I'll fix the name accordingly then. Upstream uses the `PyHamcrest` capitalisation, so I assume it makes sense to stick to that one?

Please also let me know what to do with python 3.4 (and 3.5) subports.

**Comment by pmetzger** [comment]
I think we've been mostly ignoring upstream EccentricCaps.

BTW:
> Not sure whether to remove 3.4 support from py-twisted or to add it to py-hamcrest.

I'd do the former I think.

**Comment by mojca** [comment]
I added the necessary dependencies for adding the py37 subport. The python 3.4 dependencies are deeply recursive into other ports and it seems no fun to start nailing them all (if maintainer or someone else is willing to look into this ...)

**Comment by pmetzger** [comment]
@reneeotten Any opinions?

**Comment by mojca** [comment]
> nothing in the python directory has capitals
That's not entirely true. Just by trying to auto-complete the name of PyHamcrest I get `py-Pillow`,  `py-PyQGLViewer`, `py-PyRSS2Gen`. Not saying I'm in favour of capital letter, only saying that we do have camelcasing present (and I only checked one single letter).

> On the other hand, removing py34 from py-twisted isn't that involved: only py34-scrapy depends on it...

I found the following dependencies on `py-twisted` (which did not hardcode 27) yesterday: `buildbot`, `buildbot-slave`, `coherence`, `upnp-inspector`, `splat`, `py-txgithub`, `py-structlog`, `py-nevow`, `py-axiom`, `py-epsilon`, `py-scrapy`, `py-morbid`, `alot`, `hellanzb`.
But after checking again it seems that indeed only `py-scrapy` is currently supporting 3.4 and then that one doesn't have any further dependents. @kurthindenburg, any opinion about 3.4 support in `py-scrapy`?

**Comment by mojca** [comment]
OK, other than the action about Python 3.4 in either direction, the rest should be ready. I switched to all lowercase.

Another question for @MarcusCalhoun-Lopez: since you are maintaining all other dependencies of `py-twisted`, would you be willing to take this one over as well?

**Comment by MarcusCalhoun-Lopez** [comment]
This look good to me.
> Another question for @MarcusCalhoun-Lopez: since you are maintaining all other dependencies of py-twisted, would you be willing to take this one over as well?

Yes, that would be fine.

**Comment by pmetzger** [comment]
Just an FYI, although there are a few things in the directory with capitals, most of the things with capitals upstream aren't capitalized in the directory. (At some point we might make that consistent.)

**Comment by reneeotten** [comment]
@pmetzger @mojca yeah, sorry... my statement about not having capitals in the python directory wasn't true... Nevertheless, for most of them we don't follow the upstream capitalization (for example, the py-twisted port that is also involved in this PR is upstream called Twisted).

**Comment by pmetzger** [comment]
And PySide and PySide2 are not funkycapped in our repository.

File: python/py-twisted/Portfile [hunk 1]
```diff
@@ -39,6 +39,8 @@ if {${name} ne ${subport}} {
         port:py${python.version}-incremental \
         port:py${python.version}-hyperlink \
         port:py${python.version}-constantly
+    depends_run-append \
+        port:py${python.version}-pyhamcrest
```
Comment (yan12125) [inline python/py-twisted/Portfile]: I think this should be added to depends_lib as it's listed in install_requires, so importing twisted will use PyHamcrest, too.

File: python/py-pyhamcrest/Portfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+homepage            https://github.com/hamcrest/PyHamcrest
+master_sites        pypi:P/PyHamcrest
+distname            PyHamcrest-${version}
+
+checksums           rmd160  f5fa85ec3e29c3ec41e28ae8e690700fbf2025a1 \
+                    sha256  8ffaa0a53da57e89de14ced7185ac746227a8894dbd5a3c718bf05ddbd1d56cd \
+                    size    376623
+
+python.versions     27 35 36 37
```
Comment (yan12125) [inline python/py-pyhamcrest/Portfile]: py34-twisted will break as py34-pyhamcrest is missing.

Comment (yan12125) [reply python/py-pyhamcrest/Portfile]: Sorry I didn't notice there are already discussions about this. I prefer to drop py34-scrapy, too.

**Comment by yan12125** [comment]
> There might be other dependencies missing (automat, attrs).

automat is already added to depends_run, which should be better in depends_lib. attrs should be added, too.

> Other dependencies are missing in order to be able to add py37-twisted.

Actually Twisted is not fully compatible with Python 3.7 yet. There are 4 open issues in [the upstream milestone for Python 3.7 support](https://twistedmatrix.com/trac/milestone/Python%203.7%20support).

File: python/py-twisted/Portfile (hunk 1 above)
Comment (mojca) [reply python/py-twisted/Portfile]: I don't quite understand what you are trying to say here. PyHamcrest doesn't seem to be needed during installation, so I'm not sure what we would gain by adding it to `depends_lib`. Whenever Twisted is used, PyHamcrest should be active given these rules. What am I missing?

**Comment by mojca** [comment]
> automat is already added to depends_run, which should be better in depends_lib.

I'm sorry, I apparently didn't notice that when I wrote it, but I don't understand your reasoning behind `depends_lib`.

> attrs should be added, too.

> Actually Twisted is not fully compatible with Python 3.7 yet.

I didn't check that carefully. But I would differentiate between switching to Python 3.7 as the only version being used (which could be the case in any port outside of `python` dir) and in current situation when multiple branches are offered at the same time. In the first case I would oppose switching, but in this case we are just postponing quite some manual work if you actually want to test every single `py37-something` port and add exclusively the flawless ones. Users can always use `py36-twisted`, and when support is added to a newer version of Twisted, `py37-twisted` will start working automatically.

**Comment by yan12125** [comment]
I think I got confused between depends_lib and depends_run. Sent a mail to macports-dev for discussions.

Regarding the py37-twisted subport: I consider adding a py37 subport is a signal - "twisted is ready for Python 3.7". A criteria is that existing test suite works fine. I have no strong opinion over that, though. It's fine if py37-twisted is served as "you can try to see if twisted works on Python 3.7".

**Comment by mojca** [comment]
> It's fine if py37-twisted is served as "you can try to see if twisted works on Python 3.7".

This is how I would consider/treat it. Users have a chance to test easily and report bugs in the meantime while still actively using 3.6.

**Pull request merged**